File: LLWallPaper.App.Tests/FeatureParityFixture.cs
using System.Text.Json;
using LLWallPaper.App.Models;

namespace LLWallPaper.App.Tests;

internal sealed class FeatureParityFixture
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    public SettingsFixture DefaultSettings { get; init; } = new();
    public CardFixture[] Cards { get; init; } = [];
    public SearchCaseFixture[] SearchCases { get; init; } = [];
    public RotationCaseFixture[] RotationCases { get; init; } = [];
    public BackendCaseFixture BackendCase { get; init; } = new();
    public CharacterCaseFixture[] CharacterCases { get; init; } = [];
    public SrCaseFixture[] SrCases { get; init; } = [];

    public static FeatureParityFixture Load()
    {
        var path = FindFixturePath(AppContext.BaseDirectory);
        var json = File.ReadAllText(path);
        return JsonSerializer.Deserialize<FeatureParityFixture>(json, JsonOptions)
            ?? throw new InvalidOperationException("Feature parity fixture is empty.");
    }

    private static string FindFixturePath(string startDirectory)
    {
        var directory = new DirectoryInfo(startDirectory);
        while (directory is not null)
        {
            var path = Path.Combine(
                directory.FullName,
                "tests",
                "feature-parity",
                "feature-parity.json"
            );
            if (File.Exists(path))
            {
                return path;
            }

            directory = directory.Parent;
        }

        throw new FileNotFoundException("Feature parity fixture was not found.");
    }
}

internal sealed class SettingsFixture
{
    public string BackendBaseUrl { get; init; } = "";
    public bool AutoRotateEnabled { get; init; }
    public int RotateIntervalMinutes { get; init; }
    public bool RotateOnAppStart { get; init; }
    public int RecentExcludeCount { get; init; }
    public bool PreferFavorites { get; init; }
    public bool ExcludeBlocked { get; init; }
    public bool ExcludeThirdEvolution { get; init; }
    public bool ExcludeSrCards { get; init; }
    public bool StartWithOs { get; init; }
    public bool StartMinimized { get; init; }
    public int CacheMaxMb { get; init; }
    public int HistoryMaxEntries { get; init; }
}

internal sealed class CardFixture
{
    public string Id { get; init; } = "";
    public string Name { get; init; } = "";
    public string ImageUrl { get; init; } = "";
    public string? ThumbnailUrl { get; init; }

    public CardItem ToCardItem() =>
        new()
        {
            Id = Id,
            Name = Name,
            ImageUrl = ImageUrl,
            ThumbnailUrl = ThumbnailUrl,
        };
}

internal sealed class SearchCaseFixture
{
    public string Name { get; init; } = "";
    public string Query { get; init; } = "";
    public SearchSettingsFixture Settings { get; init; } = new();
    public string[] ExpectedIds { get; init; } = [];
}

internal sealed class SearchSettingsFixture
{
    public bool ExcludeThirdEvolution { get; init; }
    public bool ExcludeSrCards { get; init; }
}

internal sealed class RotationCaseFixture
{
    public string Name { get; init; } = "";
    public RotationSettingsFixture Settings { get; init; } = new();
    public string[] FavoriteKeys { get; init; } = [];
    public string[] BlockedKeys { get; init; } = [];
    public string[] RecentKeys { get; init; } = [];
    public string? ExpectedId { get; init; }
}

internal sealed class RotationSettingsFixture
{
    public bool PreferFavorites { get; init; }
    public bool ExcludeBlocked { get; init; }
    public bool ExcludeThirdEvolution { get; init; }
    public bool ExcludeSrCards { get; init; }
}

internal sealed class BackendCaseFixture
{
    public string BaseUrl { get; init; } = "";
    public JsonElement Json { get; init; }
    public CardFixture[] ExpectedCards { get; init; } = [];
}

internal sealed class CharacterCaseFixture
{
    public string CardId { get; init; } = "";
    public string ExpectedName { get; init; } = "";
}

internal sealed class SrCaseFixture
{
    public string CardId { get; init; } = "";
    public bool Expected { get; init; }
}

File: LLWallPaper.Mac/Tests/LLWallPaperMacTests/FeatureParityFixture.swift
import Foundation
import XCTest

@testable import LLWallPaperMacCore

struct FeatureParityFixture: Decodable {
  let defaultSettings: SettingsFixture
  let cards: [CardFixture]
  let searchCases: [SearchCaseFixture]
  let rotationCases: [RotationCaseFixture]
  let backendCase: BackendCaseFixture
  let characterCases: [CharacterCaseFixture]
  let srCases: [SrCaseFixture]

  static func load(file: StaticString = #filePath, line: UInt = #line) throws
    -> FeatureParityFixture
  {
    let data = try Data(contentsOf: try fixtureUrl(file: file, line: line))
    return try JSONDecoder().decode(FeatureParityFixture.self, from: data)
  }

  static func backendJsonData(file: StaticString = #filePath, line: UInt = #line) throws -> Data {
    let data = try Data(contentsOf: try fixtureUrl(file: file, line: line))
    let root = try JSONSerialization.jsonObject(with: data)
    guard
      let dictionary = root as? [String: Any],
      let backendCase = dictionary["backendCase"] as? [String: Any],
      let json = backendCase["json"]
    else {
      XCTFail("backendCase.json was not found.", file: file, line: line)
      return Data()
    }

    return try JSONSerialization.data(withJSONObject: json)
  }

  private static func fixtureUrl(file: StaticString, line: UInt) throws -> URL {
    var directory = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)

    while true {
      let fixtureUrl =
        directory
        .appending(path: "tests")
        .appending(path: "feature-parity")
        .appending(path: "feature-parity.json")
      if FileManager.default.fileExists(atPath: fixtureUrl.path) {
        return fixtureUrl
      }

      let parent = directory.deletingLastPathComponent()
      if parent.path == directory.path {
        XCTFail("Feature parity fixture was not found.", file: file, line: line)
        throw FixtureError.notFound
      }
      directory = parent
    }
  }
}

enum FixtureError: Error {
  case notFound
}

struct SettingsFixture: Decodable {
  let backendBaseUrl: String
  let autoRotateEnabled: Bool
  let rotateIntervalMinutes: Int
  let rotateOnAppStart: Bool
  let recentExcludeCount: Int
  let preferFavorites: Bool
  let excludeBlocked: Bool
  let excludeThirdEvolution: Bool
  let excludeSrCards: Bool
  let startWithOs: Bool
  let startMinimized: Bool
  let cacheMaxMb: Int
  let historyMaxEntries: Int
}

struct CardFixture: Decodable {
  let id: String
  let name: String
  let imageUrl: URL
  let thumbnailUrl: URL?

  var cardItem: CardItem {
    CardItem(id: id, name: name, imageUrl: imageUrl, thumbnailUrl: thumbnailUrl)
  }
}

struct SearchCaseFixture: Decodable {
  let name: String
  let query: String
  let settings: SearchSettingsFixture
  let expectedIds: [String]
}

struct SearchSettingsFixture: Decodable {
  let excludeThirdEvolution: Bool
  let excludeSrCards: Bool
}

struct RotationCaseFixture: Decodable {
  let name: String
  let settings: RotationSettingsFixture
  let favoriteKeys: Set<String>
  let blockedKeys: Set<String>
  let recentKeys: Set<String>
  let expectedId: String?
}

struct RotationSettingsFixture: Decodable {
  let preferFavorites: Bool
  let excludeBlocked: Bool
  let excludeThirdEvolution: Bool
  let excludeSrCards: Bool
}

struct BackendCaseFixture: Decodable {
  let baseUrl: URL
  let expectedCards: [CardFixture]
}

struct CharacterCaseFixture: Decodable {
  let cardId: String
  let expectedName: String
}

struct SrCaseFixture: Decodable {
  let cardId: String
  let expected: Bool
}
